Common/Interfaces/IDataFileProvider.cs

```diff
@@ -0,0 +1,37 @@
+﻿/*
+ * QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
+ * Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
+ * 
+ * Licensed under the Apache License, Version 2.0 (the "License"); 
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+
+using System;
+using System.ComponentModel.Composition;
+
+namespace QuantConnect.Interfaces
+{
+    /// <summary>
+    /// Fetches a remote file for a security using a Symbol, Resolution and DateTime
+    /// Used to get remote files when they are not found locally
+    /// </summary>
+    [InheritedExport(typeof(IDataFileProvider))]
+    public interface IDataFileProvider
```

Comment:
I think it's important to document here that in order for this to work within Lean, the `IDataFileProvider` implementations *must* save the file to `Globals.DataFolder` in it's current form. Something better would be if the `Fetch` method accepts the root data directory. Accepting it as an argument is nice because it moves the dependency outside of this implementation and forces the consumer to decide where to save it.